### PR TITLE
add box_iou_rotated, ml_nms_rotated and nms_rotated

### PR DESCRIPTION
**Pull request by jiaoml1996**
add box_iou_rotated, ml_nms_rotated and nms_rotated ops for pytorch and parrots.

### CODE REVIEW

File: tests/data/config/q.py [hunk 1]
```diff
@@ -1,3 +1 @@
-custom_imports = dict(
-    imports=['r'],
-    allow_failed_imports=False)
+custom_imports = dict(imports=['r'], allow_failed_imports=False)
```
Comment (Johnson-Wang) [inline tests/data/config/q.py:1]: What is this line for?

File: mmcv/ops/box_iou_rotated.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import torch
+
+from ..utils import ext_loader
+
+ext_module = ext_loader.load_ext('_ext', ['box_iou_rotated'])
+
+
+def box_iou_rotated(bboxes1, bboxes2):
```
Comment (Johnson-Wang) [inline mmcv/ops/box_iou_rotated.py:8]: Docstring is missing.

File: mmcv/ops/nms.py [hunk 3]
```diff
@@ -302,3 +303,37 @@ def nms_match(dets, iou_threshold):
         return [dets.new_tensor(m, dtype=torch.long) for m in matched]
     else:
         return [np.array(m, dtype=np.int) for m in matched]
+
+
+def nms_rotated(boxes, iou_thr, labels=None, multi_label=False):
```
[the remark this answers is not in the document]
Comment (Johnson-Wang) [reply mmcv/ops/nms.py:308]: It seems that the arguments passed into this function is a bit different with the common ones, e.g. `softnms` and `nms`. I wonder if it would be better to define a uniform style.

**Review by Johnson-Wang — COMMENTED** [review]
Please add necessary docstrings to the ops

**Comment by Johnson-Wang** [comment]
Please kindly sign the CLA.

File: mmcv/ops/box_iou_rotated.py [hunk 4]
```diff
@@ -6,6 +6,19 @@
 
 
 def box_iou_rotated(bboxes1, bboxes2):
+    """Return intersection-over-union (Jaccard index) of boxes.
+
+    Both sets of boxes are expected to be in
+    (x_center, y_center, width, height, angle) format.
+
+    Arguments:
+        boxes1 (Tensor[N, 5])
```
Comment (Johnson-Wang) [inline mmcv/ops/box_iou_rotated.py:15]:
```python
boxes1 (Tensor): rotated bboxes 1. It has shape (N, 5), indicating (x, y, w, h, theta) for each row. 
```

File: mmcv/ops/nms.py [hunk 5]
```diff
@@ -305,7 +305,23 @@ def nms_match(dets, iou_threshold):
...
-def nms_rotated(boxes, iou_thr, labels=None, multi_label=False):
+def nms_rotated(boxes, iou_threshold, labels=None, multi_label=False):
+    """Performs non-maximum suppression (NMS) on the rotated boxes according to
+    their intersection-over-union (IoU).
+
+    Rotated NMS iteratively removes lower scoring rotated boxes which have an
+    IoU greater than iou_threshold with another (higher scoring) rotated box.
+
+    Args:
+        boxes (Tensor[N, 5]): boxes in shape (N, 5).
```
Comment (Johnson-Wang) [inline mmcv/ops/nms.py:316]:
```python
boxes (Tensor):  Rotated boxes in shape (N, 5). 
```
This comments applies to all following docstring.

File: mmcv/ops/nms.py [hunk 6]
```diff
@@ -305,7 +305,23 @@ def nms_match(dets, iou_threshold):
         return [np.array(m, dtype=np.int) for m in matched]
 
 
-def nms_rotated(boxes, iou_thr, labels=None, multi_label=False):
+def nms_rotated(boxes, iou_threshold, labels=None, multi_label=False):
```
Comment (Johnson-Wang) [inline mmcv/ops/nms.py:308]: The `score` argument is passed explicitly in `nms` and `soft_nms`. Can we support this style as well?

File: mmcv/ops/__init__.py [hunk 7]
```diff
@@ -38,5 +39,6 @@
     'RoIAlign', 'roi_align', 'RoIPool', 'roi_pool', 'SyncBatchNorm', 'Conv2d',
     'ConvTranspose2d', 'Linear', 'MaxPool2d', 'CrissCrossAttention', 'PSAMask',
     'point_sample', 'rel_roi_point_to_rel_img_point', 'SimpleRoIAlign',
-    'SAConv2d', 'TINShift', 'tin_shift'
+    'SAConv2d', 'TINShift', 'tin_shift', 'box_iou_rotated', 'ml_nms_rotated',
```
Comment (Johnson-Wang) [inline mmcv/ops/__init__.py:42]: `ml_nms_rotated` is not available any more.

Comment (Johnson-Wang) [reply mmcv/ops/__init__.py:42]: Also, please have a check over other files to avoid such errors.

File: mmcv/ops/nms.py [hunk 8]
```diff
@@ -302,3 +303,53 @@ def nms_match(dets, iou_threshold):
...
+    Returns:
+        tuple: kept dets(boxes and scores) and indice, which is always the \
+            same data type as the input.
+    """
+    if boxes.shape[0] == 0:
+        return boxes
+    if multi_label:
+        assert labels is not None
+    dets = boxes[:, :5]
+    scores = boxes[:, 5]
```
Comment (Johnson-Wang) [inline mmcv/ops/nms.py:330]: There is a bug here. The score is passed as an independent argument now, and boxes should be of shape (N, 5).

File: mmcv/ops/nms.py [hunk 9]
```diff
@@ -305,15 +305,17 @@ def nms_match(dets, iou_threshold):
...
     """Performs non-maximum suppression (NMS) on the rotated boxes according to
     their intersection-over-union (IoU).
 
     Rotated NMS iteratively removes lower scoring rotated boxes which have an
     IoU greater than iou_threshold with another (higher scoring) rotated box.
 
     Args:
-        boxes (Tensor):  Rotated boxes in shape (N, 5).
+        boxes (Tensor):  Rotated boxes in shape (N, 5). They are expected to \
+            be in (x_ctr, y_ctr, width, height, angle_degrees) format.
```
Comment (Johnson-Wang) [inline mmcv/ops/nms.py:317]: It should be angle_radian.

File: mmcv/ops/nms.py [hunk 10]
```diff
@@ -302,3 +303,53 @@ def nms_match(dets, iou_threshold):
...
+        labels (Tensor): boxes's label in shape (N,).
+        multi_label (bool): use multi label nms rotated.
+
+    Returns:
+        tuple: kept dets(boxes and scores) and indice, which is always the \
+            same data type as the input.
+    """
+    if dets.shape[0] == 0:
+        return dets, None
+    if multi_label:
```
Comment (Johnson-Wang) [inline mmcv/ops/nms.py:329]: `multi_label = labels is not None` should work, and the `multi_label` argument of this function can be thus deleted.

File: mmcv/ops/nms.py [hunk 11]
```diff
@@ -302,3 +303,53 @@ def nms_match(dets, iou_threshold):
...
+            iou_threshold=iou_threshold,
+            multi_label=multi_label)
+        keep_inds = order.masked_select(select == 1)
+        dets = dets[keep_inds, :]
+        return dets, keep_inds
+    else:
+        keep_inds = ext_module.nms_rotated(dets_wl, scores, order, dets_sorted,
+                                           iou_threshold, multi_label)
+        dets = dets[keep_inds, :]
+    return dets, keep_inds
```
Comment (Johnson-Wang) [inline mmcv/ops/nms.py:355]: Please add the following code to be compatible with other `nms` methods in this module. Note that the dets is now of shape (K, 6). 
```python
dets = torch.cat(
        (boxes[keep_inds], scores[keep_inds].reshape(-1, 1)), dim=1)
return dets, keep_inds
```

**Review by Johnson-Wang — COMMENTED** [review]
As discussed earlier, `multi_label` can be safely deleted for the purpose of clearer code structure in cuda files.

**Comment by jiaoml1996** [comment]
ok啦 @Johnson-Wang

**Review by Johnson-Wang — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]
